File: test/hardware-debug/backend/routers/serial.py
from fastapi import APIRouter, HTTPException, WebSocket, WebSocketDisconnect
import asyncio
import serial.tools.list_ports

from backend.services import modbus_service
from backend.services.log_service import save_serial_log, add_log, get_logs, clear_logs

router = APIRouter(prefix='/api/serial', tags=['串口'])

@router.get('/ports')
async def list_ports():
    ports = serial.tools.list_ports.comports()
    return [{'name': p.device, 'description': p.description} for p in ports]

@router.post('/connect')
async def connect(data: dict):
    port_name = data.get('port_name')
    baud_rate = data.get('baud_rate', 115200)
    if not port_name:
        raise HTTPException(400, '缺少串口号')
    add_log(f'正在连接 {port_name} ({baud_rate} bps)...', 'info')
    success = modbus_service.connect(port_name, baud_rate)
    if not success:
        add_log(f'{port_name} 连接失败', 'error')
        raise HTTPException(500, '串口连接失败，请检查端口号和权限')
    add_log(f'{port_name} 连接成功 ({baud_rate} bps)', 'success')
    return {'status': 'connected', 'port': port_name, 'baud_rate': baud_rate}

@router.post('/disconnect')
async def disconnect():
    modbus_service.disconnect()
    add_log('串口已断开', 'info')
    return {'status': 'disconnected'}

@router.get('/status')
async def status():
    return {
        "connected": modbus_service.is_connected(),
        "port": modbus_service.port,
        "baud_rate": modbus_service.baud_rate
    }

@router.get('/logs')
async def get_serial_logs():
    """获取通信日志"""
    return get_logs()

@router.delete('/logs')
async def clear_serial_logs():
    """清空通信日志"""
    clear_logs()
    return {'success': True}

@router.post('/logs')
async def save_log(data: dict):
    message = data.get('message', '')
    log_type = data.get('type', 'info')
    if not message:
        return {'success': False, 'message': '消息不能为空'}
    add_log(message, log_type)
    return {'success': True}

@router.websocket('/ws')
async def serial_websocket(websocket: WebSocket):
    await websocket.accept()
    try:
        while True:
            await asyncio.sleep(1)
            if modbus_service.is_connected():
                await websocket.send_json({
                    'type': 'status',
                    'connected': True,
                    'message': f'已连接 {modbus_service.port} @ {modbus_service.baud_rate}'
                })
    except WebSocketDisconnect:
        pass

File: test/hardware-debug/src/stores/serial.js
import { defineStore } from 'pinia'
import { ref, computed, onUnmounted } from 'vue'
import { api } from '../api'

export const useSerialStore = defineStore('serial', () => {
  const config = ref({
    portName: 'COM1',
    baudRate: 115200,
    dataBits: 8,
    stopBits: 1,
    parity: 'none',
    flowControl: 'none'
  })

  const connected = ref(false)
  const connecting = ref(false)
  const logs = ref([])
  const availablePorts = ref([])

  // 后端真实状态（通过轮询同步）
  const actualPort = ref(null)
  const actualBaudRate = ref(null)

  let pollTimer = null

  const statusText = computed(() => {
    if (connecting.value) return '连接中...'
    if (connected.value) return '已连接'
    return '未连接'
  })

  const statusColor = computed(() => {
    if (connecting.value) return 'amber'
    if (connected.value) return 'green'
    return 'off'
  })

  const connectionInfo = computed(() => {
    if (connecting.value) return '连接中...'
    if (connected.value) {
      // 优先使用后端返回的真实端口信息
      const port = actualPort.value || config.value.portName
      const baud = actualBaudRate.value || config.value.baudRate
      return `${port} @ ${baud}`
    }
    return '未连接'
  })

  function addLog(message, type = 'info') {
    const timestamp = new Date().toLocaleTimeString()
    logs.value.push({ timestamp, message, type })
    if (logs.value.length > 200) {
      logs.value = logs.value.slice(-200)
    }
  }

  async function fetchLogs() {
    try {
      const backendLogs = await api.getSerialLogs()
      logs.value = backendLogs
    } catch (e) {
      console.error('获取日志失败:', e)
    }
  }

  async function fetchPorts() {
    try {
      const ports = await api.getPorts()
      availablePorts.value = ports.map(p => p.name)
      if (availablePorts.value.length > 0 && !availablePorts.value.includes(config.value.portName)) {
        config.value.portName = availablePorts.value[0]
      }
    } catch (e) {
      console.error('获取串口列表失败:', e)
    }
  }

  function startPolling() {
    stopPolling()
    pollTimer = setInterval(async () => {
      try {
        const status = await api.getSerialStatus()
        // 同步后端真实连接状态
        connected.value = status.connected
        actualPort.value = status.port || null
        actualBaudRate.value = status.baud_rate || null

        // 如果后端已断开，更新配置中的端口名
        if (!status.connected && status.port) {
          config.value.portName = status.port
        }
      } catch (e) {
        // 轮询失败时不更新状态，保留上次有效值
        console.debug('串口状态轮询失败:', e.message)
      }
    }, 2000)
  }

  function stopPolling() {
    if (pollTimer) {
      clearInterval(pollTimer)
      pollTimer = null
    }
  }

  async function connect() {
    if (connecting.value || connected.value) return
    connecting.value = true
    addLog(`正在连接 ${config.value.portName}...`, 'info')

    try {
      const result = await api.connectSerial({
        port_name: config.value.portName,
        baud_rate: config.value.baudRate,
        data_bits: config.value.dataBits,
        stop_bits: config.value.stopBits,
        parity: config.value.parity
      })
      connected.value = true
      actualPort.value = result.port || config.value.portName
      actualBaudRate.value = result.baud_rate || config.value.baudRate
      addLog(`${config.value.portName} 连接成功 (${config.value.baudRate} bps)`, 'success')
    } catch (e) {
      addLog(`${config.value.portName} 连接失败：${e.message || '端口占用或不存在'}`, 'error')
    } finally {
      connecting.value = false
    }
  }

  function autoConnect(portName) {
    if (connecting.value || connected.value) return
    config.value.portName = portName
    connect()
  }

  async function disconnect() {
    try {
      await api.disconnectSerial()
    } catch (e) {
      // ignore
    }
    connected.value = false
    actualPort.value = null
    actualBaudRate.value = null
    addLog('串口已断开', 'info')
  }

  function sendData(data) {
    addLog(`发送: ${data}`, 'info')
  }

  async function clearLogs() {
    logs.value = []
    try {
      await api.clearSerialLogs()
    } catch (e) {
      // ignore
    }
  }

  return {
    config,
    connected,
    connecting,
    logs,
    availablePorts,
    actualPort,
    actualBaudRate,
    statusText,
    statusColor,
    connectionInfo,
    connect,
    autoConnect,
    disconnect,
    fetchPorts,
    fetchLogs,
    sendData,
    addLog,
    clearLogs,
    startPolling,
    stopPolling
  }
})
